crates/dtrace/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+        || unsafe { *action_data_raw }.dtrd_action as u32 == sys::DTRACEACT_EXIT
+    {
+        sys::DTRACE_CONSUME_NEXT as i32
+    } else {
+        sys::DTRACE_CONSUME_THIS as i32
+    }
+}
+
+// TODO(skywhale): Support grabbing a process for user-space inspection.
+#[derive(Debug)]
+pub struct DTrace {
+    inner: *mut sys::dtrace_hdl_t,
+    probes: Vec<ProbeData>,
+}
+
+impl DTrace {
+    pub fn new() -> Result<Self, Error> {
+        let flags = 0i32;
+        let mut code = 0i32;
+        let dtrace = unsafe { sys::dtrace_open(sys::DTRACE_VERSION as i32, flags, &mut code) };
+        if dtrace.is_null() {
+            let message_raw = unsafe { sys::dtrace_errmsg(std::ptr::null_mut(), code) };
+            return Err(Error::InitializationError(c_char_to_string(message_raw)));
+        }
+
+        Ok(Self {
+            inner: dtrace,
+            probes: Vec::default(),
+        })
+    }
+
+    /// Compiles a D program and starts collecting probe data.
+    ///
+    /// Available DTrace options:
+    /// https://docs.oracle.com/en/operating-systems/solaris/oracle-solaris/11.4/dtrace-guide/consumer-options.html
+    pub fn execute_program(&self, program: &str, options: &[(&str, &str)]) -> Result<(), Error> {
+        let program = CString::new(program).expect("CString::new failed");
+        let prog = unsafe {
+            sys::dtrace_program_strcompile(
+                self.inner,
+                program.as_ptr(),
+                sys::dtrace_probespec_DTRACE_PROBESPEC_NAME,
+                0,
+                0,
+                std::ptr::null(),
+            )
+        };
+        if prog.is_null() {
+            return Err(Error::ProgramCompilationError(self.last_error_message()));
+        }
+
+        let mut info = sys::dtrace_proginfo_t::default();
+        if unsafe { sys::dtrace_program_exec(self.inner, prog, &mut info) } == -1 {
+            return Err(Error::OperationError(self.last_error_message()));
+        }
+
+        for option in options.iter() {
+            let name = CString::new(option.0).expect("CString::new failed");
+            let value = CString::new(option.1).expect("CString::new failed");
+            if unsafe { sys::dtrace_setopt(self.inner, name.as_ptr(), value.as_ptr()) } == -1 {
+                return Err(Error::InvalidOption(self.last_error_message()));
+            }
+        }
+
+        if unsafe { sys::dtrace_go(self.inner) } != 0 {
+            return Err(Error::OperationError(self.last_error_message()));
+        }
+
+        Ok(())
+    }
+
+    // Blocks for some time and consumes the probe data collected by DTrace. The |ProgramStatus|
+    // informs about the state of the program execution.
+    //
+    // The duration it blocks depends on `switchrate`, `statusrate` and `aggrate` options used for
+    // the program execution.
+    pub fn wait_and_consume(
+        &mut self,
+        probes: &mut Vec<ProbeData>,
+    ) -> Result<ProgramStatus, Error> {
```

Comment:
Returning the `Vec` instead of taking a mutable reference may be more ergonomic?

```suggestion
    pub fn wait_and_consume(&mut self) -> Result<(ProgramStatus, Vec<ProbeData>), Error> {
```

Instead of tuple a named struct could be used. Eg.

```suggestion
    struct ConsumedData {
        status: ProgramStatus,
        data: Vec<ProbeData>
    }

    pub fn wait_and_consume(&mut self) -> Result<ConsumedData, Error>
```

though there's definitely a better name for that struct. .)